hazelcast/src/main/java/com/hazelcast/query/Predicates.java

```diff
@@ -465,6 +466,11 @@ public AndPredicate(Predicate... predicates) {
 
         @Override
         public Set<QueryableEntry> filter(QueryContext queryContext) {
+            Set<QueryableEntry> optoResult = tryOptimizedFiltering(queryContext);
+            if (optoResult != null) {
+                return optoResult;
+            }
+
             Set<QueryableEntry> smallestIndexedResult = null;
```

Comment:
It it simpler to just optimize the query? 

Predicate optimizedPredicate = optimize(predicate);

In the optimize you do a tree traversal and replace the expression. Once the predicate is optimized you execute it as it it was written using an in between instead of an or.